### DISCUSSION TITLE
Modbus CRC Check failed! wxyz!=abcd

### DISCUSSION THREAD

**Discussion by gitmarecki** [syssi/esphome-smg-ii#32] [Q&A] [answered]
I'm trying to use the same setup to figure out how to read my own inverter's log output - I have a Chinese 3.5kW unit "Raggie" with what appears to be a similar datalogging setup. I use ESPhome (2023.6.5) and tried both ESP8266 and ESP32 examples from this repo (with both development boards connected to the same MAX3232 interface). I also tried changing address of the uart device (in example we have 0x01, while my dongle says in SmartESS app that it's connected to address "5", whatever it means).

I tried multiple configurations, replacing rx with tx pins, using different pins on development boards - there is only one way it works - so I am assuming that data is flowing right. The only problem is that every time it does work, I receive no valid responses - I only get "Modbus CRC Check failed! wxyz!=abcd" messages when it reports something.

My question is - could this be a code/library/hardware issue (CRC), or does this appear that my dongle is using totally different protocol and the max3232 won't help me accessing it? Any suggestions welcome.

**Comment by syssi** [comment]
Please enable the debug output of the `uart` component and provide a log containing the raw traffic:

https://github.com/syssi/esphome-smg-ii#debugging

**Comment by gitmarecki** [comment]
[](url)
[logs_solarinverter-datalogger_logs.txt](https://github.com/syssi/esphome-smg-ii/files/12091714/logs_solarinverter-datalogger_logs.txt)

attached.

**Reply by syssi** [reply]
The responses look like garbage. Are you sure about the baud rate? Could you provide a photo of your wiring?

**Reply by gitmarecki** [reply]
it does look bad, I agree... here's the photo of the setup (I have max2323 with female connector and so is my datalogger cable, so just bridging pins 1 to 1 as in the example in the top of the readme)

connectors are done this way
MAX RX -> ESP (GPIO17) 
MAX TX -> RSP (GPIO16)

my yaml (I have no experience with uart, but I assumed tx must meet the rx on the other end)
  tx_pin: GPIO17
  rx_pin: GPIO16

![f2b92cf2-845a-4ce2-97f1-5b0d07570b95](https://github.com/syssi/esphome-smg-ii/assets/49226376/0664d6ba-8400-4fd2-91d9-746c8a1304d8)

**Reply by syssi** [reply]
To verify the full path a second photo of the inverter connector would be nice.

**Reply by gitmarecki** [reply]
I reworked the cable to ensure I am doing the same thing as in the example. The log seems to be different - I only have one set of CRC errors

[W][modbus:108]: Modbus CRC Check failed! 4810!=F880

[logs_solarinverter-datalogger_logs (1).txt](https://github.com/syssi/esphome-smg-ii/files/12093912/logs_solarinverter-datalogger_logs.1.txt)
(attaching the full file).

Also: this is how my connection interface looks like after reworking it (I previously used the proprietary one that came with the device and the wifi dongle)

![3d197ee7-df19-49ae-8f8f-6e39f5ac0551](https://github.com/syssi/esphome-smg-ii/assets/49226376/e2d0aed0-0312-40c8-b6fa-9fe4d9035636)

**Reply by syssi** [reply]
This time the ESP doesn't see any response of the inverter. There is only outgoing traffic at the log. Please try to the the Modbus address of the inverter to `1` or update the YAML to Modbuss address `0x05`. Both sides must match.

**Reply by gitmarecki** [reply, marked as answer]
after many other weird attempts, trials with different configurations (mainly rx mixed with tx on pin side and in the code), and finally at the act of desperation I got another esp (8266 this time) with another max3232 board... and by mistake I uploaded the [pipsolar](https://github.com/syssi/esphome-pipsolar/blob/main/esp8266-example.yaml) yaml... it worked without any additional configuration... I don't understand what's the difference, but happy that it worked... 
also noticed that uart speed is set to 2400 in the pipsolar repo, unlike 9600 here. Worked for me...
